### PR TITLE
implement cli feature

### PR DESCRIPTION
**Pull request by eyad-hussein**
## Describe your changes
I added the cli feature and tested it locally(did not test it on the model because i cannot install the required packages like pytorch). Please test it on your machine if you have all the necessary packages and libraries to verify it is working

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [x] If it is a core feature, I have added thorough tests.
- [x] Do we need to implement analytics?
- [x] Will this be part of a product update? If yes, please write one phrase about this update.

### CODE REVIEW

File: src/model/cfg.py [hunk 1]
```diff
@@ -1,6 +1,9 @@
 """Configurations for the model's training, loading, saving, evaluation, and Data transforms."""
 from enum import Enum
 from pathlib import Path
+import argparse
```
Comment (omar-abdelgawad) [inline src/model/cfg.py]: please reorder your imports such that the `import x`  statements are before `from x import y` statements.

File: src/model/cfg.py [hunk 2]
```diff
@@ -20,25 +23,120 @@ class DatasetType(Enum):
...
+    return ivalue
+
+def custom_arg_parser(argv: Optional[Sequence[str]] = None) -> argparse.Namespace:
+    parser = argparse.ArgumentParser(
+        description="An unsupervised image-to-image translation model",
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
+    )
+
+    parser.add_argument(
+        "-lr", "--learning_rate",
```
Comment (omar-abdelgawad) [inline src/model/cfg.py]: Short options of command-line arguments according to POSIX conventions should consist of a single character. Argparse doesn't impose this constraint, However, it could be an annoying source of buges later. For example you can apply more than one short option in a single word by concatenating characters. so `-lr` can be interpreted as applying the `-l` option and `-r` at the same time which is not true in this case (but still confusing). I recommend that you change the short option to just `-r` to refer to the `rate`.

File: src/model/cfg.py [hunk 3]
```diff
@@ -20,25 +23,120 @@ class DatasetType(Enum):
...
+        default=100,
+        help="Number of epochs",
+        metavar="NUM_EPOCHS",
+        dest="e",
+    )
+    parser.add_argument(
+        "-l", "--load_model",
+        action="store_false",
+        default=True,
+        help="Load model",
```
Comment (omar-abdelgawad) [inline src/model/cfg.py]: This is very confusing to read. The store_false is used to have a default value of True but when the argument is specefied it stores False. this means that the following line
```bash
python train.py --load_model
```
will actually NOT load the model. Can you see why this is confusing?. More importantly, We agreed that load should have a default value of False if I recall correctly. 
TLDR:
change the default to False and action to 'store_true'.

File: src/model/cfg.py [hunk 4]
```diff
@@ -20,25 +23,120 @@ class DatasetType(Enum):
...
+        dest="l",
+    )
+    parser.add_argument(
+        "-s", "--save_model",
+        action="store_true",
+        default=False,
+        help="Save model",
```
Comment (omar-abdelgawad) [inline src/model/cfg.py]: We agreed that this argument should have a default value of True (maybe you swapped load and save unintentionally). For this one maybe use the following snippet
```python
parser.add_argument(
"--no-save-model",
action = "store_false"
default=True, # This line is not needed by the way.
help = "Don't save the trained model".
dest = "SAVE_MODEL"
)
```
Note that the dest name is still save_model and not inverted. That is because the args.SAVE_MODEL will contain the True by default and we want to save by default except if we specify not to.

File: src/model/cfg.py [hunk 5]
```diff
@@ -20,25 +23,120 @@ class DatasetType(Enum):
...
+        dest="s",
+    )
+    parser.add_argument(
+        "-d", "--num_images_dataset",
+        type=positive_int,
+        default=1000,
+        help="Number of images in dataset",
```
Comment (omar-abdelgawad) [inline src/model/cfg.py]: I would change the help to `"Number of images to use from the dataset"` because that what it actually means. If the #of_images_in_dataset = x and the the cli argument --num_images_dataset = y.
The actual number of images used in training is `min(x,y)`.

File: src/model/cfg.py [hunk 6]
```diff
@@ -20,25 +23,120 @@ class DatasetType(Enum):
...
+        metavar="NUM_IMAGES_DATASET",
+        dest="d",
+    )
+    parser.add_argument(
+        "-v", "--val_batch_size",
+        type=positive_int,
+        default=8,
+        help="Validation batch size",
```
Comment (omar-abdelgawad) [inline src/model/cfg.py]: This is my mistake so I wouldn't blame you for it but batch_size name is not a good description :laughing: . It is the number of images in the grid to show in the validation step (during training). So Please update to a more elaborate help.

File: src/model/cfg.py [hunk 7]
```diff
@@ -20,25 +23,120 @@ class DatasetType(Enum):
...
+args = custom_arg_parser()
 DEVICE = "cuda" if torch.cuda.is_available() else "cpu"
-LEARNING_RATE = 2e-4
-BATCH_SIZE = 16
-NUM_WORKERS = 2
-IMAGE_SIZE = 256
+LEARNING_RATE = args.lr
+BATCH_SIZE = args.b
+NUM_WORKERS = args.w
+IMAGE_SIZE = args.i
```
Comment (omar-abdelgawad) [inline src/model/cfg.py]: Please change all dest attributes to be more descriptive. I recommend that you use the long option instead for consistency excep with special cases such as `SAVE_MODEL`. Also note that the convention for long command line arguments is to use hyphens `-` instead of underscores `_`. So please update all long options accordingly.

File: tests/custom_arg_parser_test.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    assert args.l == True
+
+def test_arg_parser_save_model():
+    assert args.s == False
+
+def test_arg_parser_num_images_dataset():
+    assert args.d == 1000
+
+def test_arg_parser_val_batch_size():
+    assert args.v == 8
```
Comment (omar-abdelgawad) [inline tests/custom_arg_parser_test.py]: Please rewrite these tests to be one named `test_default_arguments()`. Also include the `args =` line to be inside the function as each test would use a different args variable.

**Review by omar-abdelgawad — COMMENTED** [review]
Really Great Job! :partying_face: . Please review my comments and notify me as soon as you finish.
As a final note!
I noticed how much code is needed for the CLI feature and I believe that a single file should be dedicated to it. Please include all of CLI related code in a file called cli.py and then import from it in cfg.py.
```python
from model.cli import custom_arg_parser
```

**Review by omar-abdelgawad — APPROVED** [review]
LGTM!
If it FITS, it SHIPS
